php/libraries/GenericDownloader.class.inc

```diff
@@ -0,0 +1,156 @@
+     *                              the files for download.
+     *
+     * @return void
+     */
+    function __construct(string $basePath)
+    {
+        $this->realBasePath = realpath($basePath);
+        // Ensure there is a trailing forward slash.
+        $this->realBasePath = rtrim($this->realBasePath, '/') . '/';
+    }
+
+    /**
+     * Downloads a file from the filesystem.
+     *
+     * @param string $userPath The full path supplied by the calling program.
+     *                          This path can be unsanitized (i.e. can accept
+     *                          raw request data) as sanitization is done
+     *                          automatically.
+     *
+     * @return resource
+     */
+    public function downloadFile(string $userPath)
+    {
+        /* Ensure that every following test is conducted on a sanitized
+         * (non-malicious) file path.
+         */
+        $this->sanitizePath($userPath);
+        $fullPath = $userPath;
+
+        // Ensure the server is configured correctly to allow the apache user
+        // to download files from the directory where they are stored.
+        if (!is_readable($this->realBasePath)) {
+            error_log(
+                "ERROR: $this->realBasePath should be readable by the "
+                . 'apache user but is not.'
+            );
+            throw new LorisException(
+                "The server is not configured to allow downloads from the "
+                . "requested location. Please contact your administrator."
+            );
+            return;
+        }
+        // Make sure file is actually a file.
+        if (!is_file($fullPath)) {
+            error_log("ERROR: File requested for download is not a file.");
+            http_response_code(400);
+            return;
+        }
+
+        // Make sure the file itself is readable.
+        if (!is_readable($fullPath)) {
+            error_log("ERROR: A forbidden file was requested for download");
+            http_response_code(403);
+            return;
+        }
+
+        // Make sure the requested file exists.
+        if (!file_exists($fullPath)) {
+            error_log("ERROR: File requested for download does not exist.");
+            http_response_code(404);
+            return;
+        }
+
+        $stream = new \Zend\Diactoros\Stream($fullPath);
+        return $stream->detach();
+    }
+
+    /**
+     * Returns the MIME type given a file
+     *
+     * @param string $fullPath The path to the file.
+     *
+     * @return string The MIME type of $filename.
+     */
+    function getContentType(string $fullPath): string
+    {
+        if (!is_file($fullPath)) {
+            $pathForPrinting = htmlentities($fullPath);
+            throw new LorisException(
+                "Cannot get mimetype of $pathForPrinting because it is not a "
+                . "file!"
+            );
+            return '';
+        }
+        return mime_content_type($fullPath);
```

Comment:
This should also use sanitized path. If not, it could be used to query and probe for existence of specific file

Reply:
still the line 126 is not using the sanitized path.

Reply:
Nope, it will raise an exception if the file doesn't exist and log an error if the file contain ".." but will no break in that case

Reply:
Hmm yeah you're right it works differently after I made the changes to allow for sym links. This needs more work